src/chat.rs

```diff
@@ -1291,6 +1295,75 @@ impl ChatId {
             .unwrap_or_default();
         Ok(protection_status)
     }
+
+    /// Returns the sort timestamp for a protection message.
+    ///
+    /// `timestamp` should be normally the "sent" timestamp of the message caused the chat
+    /// protection state change.
+    pub(crate) async fn calc_protection_msg_sort_timestamp(
+        self,
+        context: &Context,
+        timestamp: i64,
+    ) -> Result<i64> {
+        let sort_to_bottom = true;
+        self.calc_sort_timestamp(context, timestamp, sort_to_bottom, false)
+            .await
+    }
+
+    /// Returns the sort timestamp for a new message in the chat.
+    ///
+    /// `message_timestamp` should be either the message "sent" timestamp or a timestamp of the
+    /// corresponding event in case of a system message (usually the current system time).
+    /// `always_sort_to_bottom` makes this ajust the returned timestamp up so that the message goes
+    /// to the chat bottom.
+    /// `incoming` -- whether the message is incoming.
+    pub(crate) async fn calc_sort_timestamp(
+        self,
+        context: &Context,
+        message_timestamp: i64,
+        always_sort_to_bottom: bool,
+        incoming: bool,
+    ) -> Result<i64> {
+        let mut sort_timestamp = cmp::min(message_timestamp, smeared_time(context));
+
+        let last_msg_time: Option<i64> = if always_sort_to_bottom {
+            // get newest message for this chat
+            context
+                .sql
+                .query_get_value(
+                    "SELECT MAX(timestamp) FROM msgs WHERE chat_id=? AND state!=?",
+                    (self, MessageState::OutDraft),
```

Comment:
Intuitively, here we should also check whether it's a hidden message, not only whether it's a draft

> But let hidden messages sort together with protection messages because hidden messages also can be or not be verified.

But hidden messages aren't shown to the user, and the sort timestamp is about in which order messages are shown to the user?

Reply:
When it comes to protection messages, the sort timestamp is not only about the order of messages shown to the user, but also about which messages are verified. Hidden messages such as location ones also can be verified. If we allow them to have a relaxed ordering with protection messages, we lose information about their verification status. This information can be useful to display some additional info in the chat, maybe not now but in the future DC versions, so i'd like to preserve it

Reply:
Or i'm wrong and `Param::GuaranteeE2ee` is for that?

Reply:
But as i see from the code, a message can be encrypted (and displayed with a padlock), but not verified, even in verified chats. Also checked this in my DC Desktop

Reply:
> Or i'm wrong and `Param::GuaranteeE2ee` is for that?

No it's not

> But as i see from the code, a message can be encrypted (and displayed with a padlock), but not verified (be in an unverified section), even in verified chats. Also checked this in my DC Desktop

Yes

----

> This information can be useful to display some additional info in the chat, maybe not now but in the future DC versions, so i'd like to preserve it

Usually when one tries to plan for a possible future like this, it never happens.

That said, it also isn't bad as it is right now - it violates the principle of least astonishment a little bit because here we only check for whether it's a draft and a bit further down we also check for whether it's hidden, and the reason for that isn't clear for the reader, but that's all.

Reply:
Added a comment on this so that it is easier to understand for readers